design/plankton/CommunicationLayerDesign.md

```diff
@@ -36,13 +36,161 @@ The IBEX team at ISIS, for example, have expressed interest in using Plankton. N
+
+
+
+#### Environment & Timing Loop
+
+The design separates the timing loop from the `Adapter`. The while loop is built into an `Environment`, of which there is just one per simulation. We would like to be able to make updates to the `Environment` without having ot restart the simulation or renew the adapters. One possible implementation of the `Environment` might look like this:
+
+```python
+
+class Environment(object):
+
+    def __init__(self, adapter):
+        self._adapter = adapter # adapter wraps device (see above)
+        self._count = 0
+        self._timer = 0.0
+        self._delta = 0.0
+        self._user_delta = None
+        
+    def _delta(self):
+        if self._user_delta:
+            return self._user_delta
+        else:
+            return (datetime.now() - start).total_seconds()
+            
+    def set_user_delta(self, delta):
+        self._user_delta = delta
+
+    def run(self):
+
+        while True:
+            start = datetime.now()
+
+            
+            self._adapter.process(delta) # stream adapters would call asyncore.loop(0.1, count=1) as part of this
+
+            
+            count += 1
+            timer += self._delta()
+            self._delta = self._delta()
+            # ....
+```
+
+#### Environment Updates
+
+The separation of the runnable parts of the simulation (now in the `Environment`) from the `Device` and the `Adapter` parts should make updates which affect only the `Environment` easier.
+
+Here's one possibility I've been considering:
+
+`EnvironmentAdapter` sees `Environment` 
+
+```python 
+    class Environment(object):
+
+    def __init__(self, adapter, environ_adapter):
+        self._environ_adapter = environ_adapter
+        self._environ_adapter.set_adaptee(self)
+        self._adapter = adapter
+
+    def run(self):
+
+        while True:
+            # Everything still driven from a central processing loop.
+            self._environment_adapter.process()
+            
+            self._adapter.process(delta)
+
+```
+
+```python
+   class StreamEnvironmentAdapter(object):
+
+    def set_adaptee(self, environment):
+        self._environment = environment
+        
+    
+    def process(self):
+        asyncore.loop(0.1, count=1)
+        # ...
+        self._environment.set_user_delta(user_value)
+        
+```
+
+After discussion with Michael Wedel, I am now strongly leaning towards providing a simple single protocol for the `Environment Control`. While `Adapter` flexibility and variation is vital in order to support different device types. the `Environment` is internal to plankton and we should publish and control a single mechanism for this. Michael Wedel has suggested JSON over zeromq.
```

Comment:
> If the simulation and all its aspects are run in one thread and the "command server" in another, does it affect the determinism of the simulation?

Theoretically, yes... in practice, I can't think of a scenario where it would matter (which doesn't mean there isn't one).

By introducing a separate thread for the command server, we can no longer be sure when exactly (on which cycle) the command takes effect, if the command is issued while the simulation is right on the edge between two cycles. 

However, when running a simulation normally, the determinism aspect isn't that important... after all, there are many non-deterministic factors acting on the simulation from outside (not least of which the user themselves). And when running a unit test, you're doing everything sequentially: you would call `process` yourself and, I assume, you'd also issue commands directly and not via the command server (so not in a separate thread).